### ISSUE TITLE
Backend Plugins (for Auth, TSDB, etc) 

### ISSUE THREAD

**Issue by torkelo** [grafana/grafana#6841]
Grafana needs a way to extend backend features. 

Backend plugin support & api
- for auth, like Keystone, better LDAP etc. 
- support for more TSDBs for alerting.

**Comment by fg2it** [comment]
@torkelo I suggest to also move phantomjs as a backend plugin. This would be an other step towards #5156. It would ease an official release for arm, the phantomjs plugin being only distributed for x86_64.

**Comment by fg2it** [comment]
Yes, no doubt on this on my side.
But, is it a fundamental feature (in the sense grafana would not work without it) ? What would be the down side to have it in a separate plugin ? It could be release in a separate .deb or rpm so install would be as easy as it is to install grafana.

**Comment by torkelo** [comment]
@fg2it yes, well I am open to any PR that makes the phantomjs dependency optional

**Comment by raiseking** [comment]
@torkelo , what is the progress of the backend plugin develop for cloud foundry, thanks!

**Comment by torkelo** [comment]
no progress yet, may be able to start on this after summer

**Comment by PhilSmithABB** [comment]
Will it be possible to develop custom backend plugin's ? We have our own proprietary tsdb and it would be great if we could setup alerts and notifications for it.

**Comment by torkelo** [comment]
yes, that is the main incentive to enable backend plugins, to enable custom backend plugins

**Comment by bergquist** [comment]
Been doing some tests with https://github.com/hashicorp/go-plugin and I like it. The fact that we can use GRPC and therefore make plugins language agnostic is really powerful. The performance lose of using grpc shouldnt matter for any of the use cases in grafana.

**Comment by bergquist** [comment]
work branch https://github.com/grafana/grafana/tree/backend_plugins

**Comment by bergquist** [comment]
Some.. shattered updates about this issue.

> TL;DR. We 


### Tech decisions

#### Plugin framework
Backend plugins will be implemented using Hashicorps go-plugin library. This gives us a few advantages over the plugin lib in the standard library. Hashicorps go-plugin supports windows, it's a subprocess so a plugin cannot modify or crash Grafana itself, Hashicorps go-plugin support sending data over gRPC which makes it possible to write plugins in any language, not just Go. The only downside with Hashicorps go-plugin compared to stdlib is performance, but it's fast enough for our use-cases. 

#### Extracting contracts to separate repos.
Currently, all proto files and generated gRPC structs are bundled within Grafana. I'm not sure if we should keep it that way or if we should extract it to something like `grafana/grafana-models`. Regardless, the proto files and generated files should be versionable by `dep, godep, govendor` etc. 

#### plugin.json
We have introduced two new properties for `plugin.json`

```json
{
  "backend" : true,
  "executable": "simple-plugin"
}
```
#### All data sources should use plugin protobuf structs for req/res
Once the development of backend plugins is more stable, all internal data sources should use the generated request/response structs. Since that change would touch so many areas we introduced a wrapper that can be used until the backend plugins matured a bit more. 

### Plugin order
- Data sources
- Alert notifiers
- Authentication/Authorization
- External image uploader

### Todo before we merge to master

- [ ] Review proto contracts. 
- [ ] Current implementation is quite `POCCY`, it should be hardened.
- [ ] Decide how we should manage contracts (separate repo or not)
- [ ] Build at least one golang plugin and one python plugin.
- [ ] Start planning for how we should manage plugin distribution. 

work branch: https://github.com/grafana/grafana/tree/backend_plugins
plugin POC: https://github.com/bergquist/simple-json-datasource/tree/backend
we use Hashicorps plugin lib for loading plugins in the backend. https://github.com/hashicorp/go-plugin

**Comment by ASimpson87** [comment]
100% Want this feature to enable running in NodeJS env. Will be awesome!

**Comment by h7kanna** [comment]
Hi team,
Can you please update the state of backend plugin support?

Thanks,

**Comment by andig** [comment]
> support for more TSDBs for alerting.

ping @bergquist for an update. Would *love* alerting with simple-json.

**Comment by Cerfoglg** [comment]
Pinging @bergquist for more updates on backend plugins support. We would love to be able to use alerts with a CrateDB Datasource

**Comment by Gauravshah** [comment]
@torkelo Unable to find the branch https://github.com/grafana/grafana/tree/backend_plugins also it seems it was not merged. Is there a way to contribute ?
cc @bergquist

**Comment by Gauravshah** [comment]
sorry my bad the branch does seem to be merged. https://github.com/grafana/grafana/commit/3951df136c65ed70f06b7a4d9213636c53eb1feb Thanks will take a look and use it

**Comment by Gauravshah** [comment]
@bergquist tried using  https://github.com/grafana/simple-json-datasource/tree/backend-updated but it does not work well with alerting. Even if we set the window as 5m the data being returned is for the whole data source ( end to end 1 day or so) . Is it a limitation on the grafana side or plugin side ?

**Comment by sbueringer** [comment]
Hi,
I've wrote & published to grafana consul datasource plugin. There seem's to be a problem with the grafana cli when installing it (https://github.com/sbueringer/grafana-consul-datasource/issues/1, https://github.com/sbueringer/grafana-consul-datasource/issues/2). The backend binary is not made executable. 

https://github.com/grafana/grafana-plugin-repository/pull/305#issuecomment-425681339:
> This plugin is using an alpha feature of Grafana - backend plugins. The backend plugin works fine. The problem is with the distribution of the plugin which we have not worked on or solved yet. For example, when I installed this plugin I had to set the permissions of the grafana-consul-plugin_linux_amd64 manually to get it to work. Not sure what to do about this - will discuss it with the team.

Should I create a new issue for this or is it already getting fixed?

**Comment by bergquist** [comment]
I'm not sure we can expect Grafana to be able to make binaries executable. I think it would be preferred if the shipped binary is executable. But I'm not a security expert :) So any advice is welcome.

**Comment by sbueringer** [comment]
I'm also no security expert :). So if I understand the plugin publishing & installation process correctly, it works like this:
1. Create a plugin in a Git Repository and upload the already built plugin in the dist folder
2. Get a PR in grafana-plugin-repository merged with a reference to a commit of your GIT repository
3. The publishing process then packages the dist folder from your repository and publishes it
4. The grafana-cli installs the plugin via downloading & extracting the plugin.

So in my case the backend binary after step 2 is uploaded with a executable flag. I think somewhere in step 3 or 4 the executable bit vanishes :). So I think the problem must be either addressed in step 3 or 4 or the user has to manually make the plugin executable, which wouldn't be a great UX.

**Comment by Gauravshah** [comment]
Hi,

we are trying to move https://github.com/GoshPosh/grafana-meta-queries a `mixed` datasource that helps do arithmetic on top existing queries.
currently in frontend plugin we have
```
function MetaQueriesDatasource($q, datasourceSrv) {
this.query = function(options) {
...
options.targets
```
so using `datasourceSrv` we can access all of the `options.targets`

But in backend we have access to :
```
func (t *DruidDatasource) Query(ctx context.Context, tsdbReq *datasource.DatasourceRequest) (*datasource.DatasourceResponse, error) {
```
which gives us access only to the current datasource. Assuming that all of the datasources that metaqueries is going to use is also backend we should be able to query grafana. Any directions on how should we move forward ?

we have successfully moved https://github.com/GoshPosh/druidplugin/tree/druid_alert to backend.

**Comment by Gauravshah** [comment]
@bergquist can you help with some direction for mixed datasource plugin ?

**Comment by ShilpaSivanesan** [comment]
+1 @bergquist could you suggest on getting alerting work with mixed data source plugin

**Comment by ShilpaSivanesan** [comment]
> Hi,
> 
> we are trying to move https://github.com/GoshPosh/grafana-meta-queries a `mixed` datasource that helps do arithmetic on top existing queries.
> currently in frontend plugin we have
> 
> ```
> function MetaQueriesDatasource($q, datasourceSrv) {
> this.query = function(options) {
> ...
> options.targets
> ```
> so using `datasourceSrv` we can access all of the `options.targets`
> 
> But in backend we have access to :
> 
> ```
> func (t *DruidDatasource) Query(ctx context.Context, tsdbReq *datasource.DatasourceRequest) (*datasource.DatasourceResponse, error) {
> ```
> which gives us access only to the current datasource. Assuming that all of the datasources that metaqueries is going to use is also backend we should be able to query grafana. Any directions on how should we move forward ?
> 
> we have successfully moved https://github.com/GoshPosh/druidplugin/tree/druid_alert to backend.

+1 @bergquist could you suggest on getting alerting work with mixed data source plugin

**Comment by sbueringer** [comment]
@Gauravshah Afaik you only have access to your current datasource, which explicitly send to the backend via jsonData or secureJsonData (https://github.com/sbueringer/grafana-consul-datasource/blob/master/datasource.go#L356-L371)

**Comment by lcoulet** [comment]
Hello, this is a feature we'd really like to see coming (support of alerting & notifications), because we have a custom plugin to a TSDB.  Is there anything we can do to help - e.g. by contributing?

**Comment by tuckson** [comment]
Anything on this? Am so waiting for the MultiQuery Datasource to have to possibility to alert on.

**Comment by umniktech** [comment]
Hi All, 

Just want to make sure that Zabbix-Grafana integration still not working for alert setup in Grafana?

**Comment by deanet** [comment]
> Hi All,
> 
> Just want to make sure that Zabbix-Grafana integration still not working for alert setup in Grafana?

"There has been no big PRs & community enhancents to alerting in the mean time so the feature has not improved much since it’s inception" - The Author said :D 

Ref: https://community.grafana.com/t/ok-thats-enough-for-me/9203/7

**Comment by llKetanll** [comment]
2 years... Is this going to happen?

**Comment by OptrixAU** [comment]
Yeah - I've got a data source that I would very much like to allow alerting on.

For alerting specifically, why should we **have** to make alerting plugins? Surely if the alert is based on a query, Grafana could schedule the query in the background and run it regularly for a very short time-frame to see if the alert is active. 

Heck, it could even adjust the query time so that you can specify in the alert how long the data has to be above the level before the alert triggers.

Sure, plugins can make it much more efficient...but surely someone could make a generic solution for alerts without too much trouble?

**Comment by hagen1778** [comment]
@OptrixAU The alert requests are performed on backend which is written in Go. The backend has no idea how to form and send requests for each particular datasource, it also doesn't know how to parse the response.  So in order to get alerting for any plugin someone needs to implement the same request sending/parsing logic in Go.

**Comment by OptrixAU** [comment]
I haven't set up any alerts as yet, but from what I've read the idea is that you create alerts based on a chart - choosing where the activation threshold is, plus other factors such as the time it needs to remain over that threshold, hysterisis etc.

To make a 'generic' alerter, can't you just use this information to generate the routine query? You already know the Data Source that needs to be queried, plus all of the details of the query itself. Surely in GO you can send the query request, get the time series back, and examine the series to find if the alert condition is true?

The only difference between the original query used to create the alert and the **actual** query used to check for the alert is the time window. When performing your routine alert check, the time window should be 'now - <minimum-alert-time>' - so if you're making an alert that triggers if CPU usage has been high for more than 30 seconds, you'd use a **now-30s** time window.

**Comment by torkelo** [comment]
Not sure what your trying to say

**Comment by agnivade** [comment]
Hi, according to this [comment](https://github.com/grafana/grafana/issues/6818#issuecomment-361158504), it seems that Grafana won't be adding any additional notification channels and writing new backend plugins is the only way forward.

That is sad because I was looking forward to integrate Mattermost with Grafana. Since this issue is not gaining any more traction, where do you suggest is the best place to document the fact that Mattermost is compatible with the Slack format ? 

Perhaps adding a section for Mattermost here https://github.com/grafana/grafana/blob/master/docs/sources/alerting/notifications.md is a good idea ?

**Comment by bergquist** [comment]
@agnivade this issue is getting lots of attention internally. I can't say when we will support alert notifiers and plugins but backend plugins are currently getting a big overhaul.

**Comment by witekest** [comment]
> plugins but backend plugins are currently getting a big overhaul.

Do you have authentication plugins on the radar as well?

**Comment by agnivade** [comment]
@bergquist - That's fine. Can you let me know where is the best place to document that Mattermost is compatible with the Slack format ? Essentially, I am trying to get an answer to my second question in the original post.

**Comment by AlienHoboken** [comment]
@bergquist Is there anything an external contributor can do to further the effort? My organization currently uses a webhook notification channel as when we built our integration (Aug. 2019) this issue appeared to be de-prioritized.

Would love to use the new backend plugin system or even provide some feedback on / test an early version.

**Comment by Koojav** [comment]
@bergquist Does this: https://github.com/grafana/grafana/issues/22511 mean that in Grafana 7.0 it will be possible to set up alerts based on BigQuery data source?

**Comment by bergquist** [comment]
@Koojav Yes. If the BigQuery plugin is implemented as a backend plugin.

**Comment by paulogcrio** [comment]
Any predictions to correct the alert problem? 
Grafana Version 7
Pluguin Zabbix Version 3.11.0
Zabbix Version 4.4.7
![image](https://user-images.githubusercontent.com/50876761/79152517-bac70300-7da2-11ea-85af-35cfdbacd93a.png)

![image](https://user-images.githubusercontent.com/50876761/79153059-a0415980-7da3-11ea-8f56-6b0bab630399.png)

**Comment by Koojav** [comment]
> @Koojav Yes. If the BigQuery plugin is implemented as a backend plugin.

@bergquist and this: https://grafana.com/grafana/plugins/doitintl-bigquery-datasource is a **backend** plugin, right?

**Comment by bergquist** [comment]
The tech is done and we are now working on docs tutorials for building datasource backend plugins. 

ref https://github.com/grafana/grafana/issues/16066 for docs.

@Koojav its needs a alerting flag enables in the plugin.json as well.

**Comment by billylongta** [comment]
Hi @bergquist how about the release plan for the alert feature? are there any ways for creating an alert while waiting for release this one?
